src/assembly/manifest.ts

```diff
@@ -0,0 +1,166 @@
+import { AssetManifest, DockerImageManifestEntry, FileManifestEntry } from 'cdk-assets';
+import { StackManifest } from './stack';
+import { ConstructTree, StackAsset, StackMetadata } from './types';
+
+/**
+ * Reads a Cloud Assembly manifest
+ */
+export class AssemblyManifestReader {
+    public static readonly DEFAULT_FILENAME = 'manifest.json';
+
+    /**
+     * Reads a Cloud Assembly manifest from a file
+     */
+    public static fromFile(fileName: string): AssemblyManifestReader {
+        try {
+            const obj = Manifest.loadAssemblyManifest(fileName);
+            return new AssemblyManifestReader(path.dirname(fileName), obj);
+        } catch (e: any) {
+            throw new Error(`Cannot read manifest '${fileName}': ${e.message}`);
+        }
+    }
+
+    /**
+     * Reads a Cloud Assembly manifest from a file or a directory
+     * If the given filePath is a directory then it will look for
+     * a file within the directory with the DEFAULT_FILENAME
+     */
+    public static fromPath(filePath: string): AssemblyManifestReader {
+        let st;
+        try {
+            st = fs.statSync(filePath);
+        } catch (e: any) {
+            throw new Error(`Cannot read manifest at '${filePath}': ${e.message}`);
+        }
+        if (st.isDirectory()) {
+            return AssemblyManifestReader.fromFile(path.join(filePath, AssemblyManifestReader.DEFAULT_FILENAME));
+        }
+        return AssemblyManifestReader.fromFile(filePath);
+    }
+
+    /**
+     * The directory where the manifest was found
+     */
+    public readonly directory: string;
+
+    private readonly assembly: CloudAssembly;
+    private readonly stacks = new Map<string, CloudFormationStackArtifact>();
+    private readonly _stackManifests = new Map<string, StackManifest>();
+    private readonly tree: ConstructTree;
+
+    constructor(directory: string, private readonly manifest: AssemblyManifest) {
+        this.directory = directory;
+        this.assembly = new CloudAssembly(directory, {
+            // we don't need version checking / version checking would mean we would have to
+            // publish a new version of the library everytime the cdk version increases
+            skipVersionCheck: true,
+        });
+        this.tree = fs.readJsonSync(path.resolve(this.directory, 'tree.json')).tree;
+        if (!this.tree.children) {
+            throw new Error('Invalid tree.json found');
+        }
+        this.renderStackManifest();
+    }
+
+    private renderStackManifest() {
+        for (const [artifactId, artifact] of Object.entries(this.manifest.artifacts ?? {})) {
+            if (artifact.type === ArtifactType.AWS_CLOUDFORMATION_STACK) {
+                const stackArtifact = this.assembly.getStackArtifact(artifactId);
+                this.stacks.set(artifactId, stackArtifact);
+                const metadata: StackMetadata = {};
+                const props = artifact.properties as AwsCloudFormationStackProperties;
```

Comment:
Are we sure that type is compatible? Is there any way we can check at runtime and gracefully fail otherwise?

Reply:
I'm sure, but since we only need one property from it I can do this without type casting.